spec/requests/users_controller_spec.rb

```diff
@@ -1,51 +1,2402 @@
+
+        describe 'invalid 2 factor token' do
+          it 'should display the right error' do
+            second_factor
+
+            put "/u/admin-login/#{email_token.token}", params: { second_factor_token: '13213' }
+
+            expect(response.status).to eq(200)
+            expect(response.body).to include(I18n.t('login.second_factor_description'));
+            expect(response.body).to include(I18n.t('login.invalid_second_factor_code'));
+          end
+        end
+
+        it 'logs in when a valid 2-factor token is given' do
+          put "/u/admin-login/#{email_token.token}", params: { second_factor_token: ROTP::TOTP.new(second_factor.data).now }
+
+          expect(response).to redirect_to('/')
+          expect(session[:current_user_id]).to eq(admin.id)
+        end
+      end
+    end
+  end
+
+  describe '#toggle_anon' do
+    it 'allows you to toggle anon if enabled' do
+      SiteSetting.allow_anonymous_posting = true
+
+      user = sign_in(Fabricate(:user))
+      user.trust_level = 1
+      user.save!
+
+      post "/u/toggle-anon.json"
+      expect(response).to be_success
+      expect(session[:current_user_id]).to eq(AnonymousShadowCreator.get(user).id)
+
+      post "/u/toggle-anon.json"
+      expect(response).to be_success
+      expect(session[:current_user_id]).to eq(user.id)
+    end
+  end
+
+  describe '#create' do
+    def honeypot_magic(params)
+      get '/u/hp.json'
+      json = JSON.parse(response.body)
+      params[:password_confirmation] = json["value"]
+      params[:challenge] = json["challenge"].reverse
+      params
+    end
+
+    before do
+      UsersController.any_instance.stubs(:honeypot_value).returns(nil)
+      UsersController.any_instance.stubs(:challenge_value).returns(nil)
+      SiteSetting.allow_new_registrations = true
+      @user = Fabricate.build(:user)
+      @user.password = "strongpassword"
+    end
+
+    let(:post_user_params) do
+      { name: @user.name,
+        username: @user.username,
+        password: "strongpassword",
+        email: @user.email }
+    end
+
+    def post_user
+      post "/u.json", params: post_user_params
+    end
+
+    context 'when email params is missing' do
+      it 'should raise the right error' do
+        post "/u.json", params: {
+          name: @user.name,
+          username: @user.username,
+          passsword: 'tesing12352343'
+        }
+        expect(response.status).to eq(400)
+      end
+    end
+
+    context 'when creating a user' do
+      it 'sets the user locale to I18n.locale' do
+        SiteSetting.default_locale = 'en'
+        I18n.stubs(:locale).returns(:fr)
+        post_user
+        expect(User.find_by(username: @user.username).locale).to eq('fr')
+      end
+    end
+
+    context 'when creating a non active user (unconfirmed email)' do
+      it 'returns a 500 when local logins are disabled' do
+        SiteSetting.enable_local_logins = false
+        post_user
+
+        expect(response.status).to eq(500)
```

Comment:
Hmm we should fix this to return a proper response status instead of a 500. Any error can trigger a 500 error here.